File: backend/models/scenario.py
from pydantic import BaseModel


class ReturnProfile(BaseModel):
    mean_pct: float
    stddev_pct: float


class InvestmentReturns(BaseModel):
    stocks_mean_pct: float = 8.0
    stocks_stddev_pct: float = 16.0
    bonds_mean_pct: float = 4.0
    bonds_stddev_pct: float = 6.0
    real_estate_appreciation_pct: float = 3.5


class Inflation(BaseModel):
    general_mean_pct: float = 3.0
    general_stddev_pct: float = 1.0
    college_tuition_pct: float = 5.0
    healthcare_pct: float = 6.0


class AssetAllocation(BaseModel):
    stocks_pct: float = 70
    bonds_pct: float = 25
    cash_pct: float = 5


class AllocationStrategy(BaseModel):
    pre_retirement: AssetAllocation = AssetAllocation()
    post_retirement: AssetAllocation = AssetAllocation(stocks_pct=50, bonds_pct=40, cash_pct=10)
    glide_path_start_years_before: int = 5


class CollegeAssumptions(BaseModel):
    annual_cost_today: float = 65000
    room_and_board_today: float = 18000
    financial_aid_annual: float = 0
    scholarship_annual: float = 0


class SocialSecurityAssumptions(BaseModel):
    primary_pia_at_67: float = 3200
    spouse_pia_at_67: float = 1800
    claiming_age_primary: int = 67
    claiming_age_spouse: int = 67
    cola_pct: float = 2.0


class HealthcareAssumptions(BaseModel):
    annual_premium_today: float = 24000
    annual_out_of_pocket_today: float = 6000
    pre_medicare_gap_years: int = 2
    aca_marketplace_annual: float = 30000
    medicare_annual: float = 8000


class LargePurchase(BaseModel):
    name: str
    year: int
    purchase_price: float = 0
    down_payment_pct: float = 25
    mortgage_rate_pct: float = 6.5
    mortgage_term_years: int = 30
    # Annual carrying costs (property tax, insurance, maintenance, HOA, utilities)
    annual_carrying_cost: float = 0
    annual_property_tax: float = 0
    # Rental conversion fields
    is_rental_conversion: bool = False
    conversion_cost: float = 0
    monthly_rental_income: float = 0
    vacancy_rate_pct: float = 8
    annual_maintenance_pct: float = 1.0
    property_management_pct: float = 10
    current_mortgage_balance: float = 0
    current_mortgage_payment: float = 0


class LifeEvent(BaseModel):
    """One-time cash event (inheritance, windfall, major expense, etc.)."""
    name: str
    year: int
    amount: float = 0  # positive = inflow to portfolio, negative = outflow
    taxable: bool = False  # if True, apply income tax rate
    tax_rate_override: float | None = None  # custom rate (e.g. estate tax), None = use profile rate


class Assumptions(BaseModel):
    investment_returns: InvestmentReturns = InvestmentReturns()
    inflation: Inflation = Inflation()
    asset_allocation: AllocationStrategy = AllocationStrategy()
    college: CollegeAssumptions = CollegeAssumptions()
    social_security: SocialSecurityAssumptions = SocialSecurityAssumptions()
    healthcare: HealthcareAssumptions = HealthcareAssumptions()
    large_purchases: list[LargePurchase] = []
    life_events: list[LifeEvent] = []
    return_profiles: dict[str, ReturnProfile] = {}


class Scenario(BaseModel):
    schema_version: int = 1
    name: str
    description: str = ""
    assumptions: Assumptions = Assumptions()

File: frontend/src/types/scenario.ts
export interface ReturnProfile {
  mean_pct: number;
  stddev_pct: number;
}

export interface InvestmentReturns {
  stocks_mean_pct: number;
  stocks_stddev_pct: number;
  bonds_mean_pct: number;
  bonds_stddev_pct: number;
  real_estate_appreciation_pct: number;
}

export interface Inflation {
  general_mean_pct: number;
  general_stddev_pct: number;
  college_tuition_pct: number;
  healthcare_pct: number;
}

export interface AssetAllocation {
  stocks_pct: number;
  bonds_pct: number;
  cash_pct: number;
}

export interface AllocationStrategy {
  pre_retirement: AssetAllocation;
  post_retirement: AssetAllocation;
  glide_path_start_years_before: number;
}

export interface CollegeAssumptions {
  annual_cost_today: number;
  room_and_board_today: number;
  financial_aid_annual: number;
  scholarship_annual: number;
}

export interface SocialSecurityAssumptions {
  primary_pia_at_67: number;
  spouse_pia_at_67: number;
  claiming_age_primary: number;
  claiming_age_spouse: number;
  cola_pct: number;
}

export interface HealthcareAssumptions {
  annual_premium_today: number;
  annual_out_of_pocket_today: number;
  pre_medicare_gap_years: number;
  aca_marketplace_annual: number;
  medicare_annual: number;
}

export interface LargePurchase {
  name: string;
  year: number;
  purchase_price: number;
  down_payment_pct: number;
  mortgage_rate_pct: number;
  mortgage_term_years: number;
  annual_carrying_cost: number;
  annual_property_tax: number;
  is_rental_conversion: boolean;
  conversion_cost: number;
  monthly_rental_income: number;
  vacancy_rate_pct: number;
  annual_maintenance_pct: number;
  property_management_pct: number;
  current_mortgage_balance: number;
  current_mortgage_payment: number;
}

export interface LifeEvent {
  name: string;
  year: number;
  amount: number;
  taxable: boolean;
  tax_rate_override: number | null;
}

export interface Assumptions {
  investment_returns: InvestmentReturns;
  inflation: Inflation;
  asset_allocation: AllocationStrategy;
  college: CollegeAssumptions;
  social_security: SocialSecurityAssumptions;
  healthcare: HealthcareAssumptions;
  large_purchases: LargePurchase[];
  life_events: LifeEvent[];
  return_profiles: Record<string, ReturnProfile>;
}

export interface Scenario {
  schema_version: number;
  name: string;
  description: string;
  assumptions: Assumptions;
}

export interface ScenarioListItem {
  slug: string;
  name: string;
}
